### DISCUSSION TITLE
[Question][Hw3] Segmentation Fault from __coverage__

### DISCUSSION THREAD

**Discussion by adilb99** [prosyslab-classroom/cs348-information-security#523] [Q&A]
When I use `__coverage__` function from _runtime.c_ , it produces a segmentation fault.

I tried checking it with `if(f == NULL) fprintf(stdout, "IT'S NULL\n");` and it does print.

Therefore, the line `fprintf(f, "%d,%d\n", line, col);` causes a segfault due to the `f` file pointer being NULL.

Why does this happen?

**Comment by RiceBiscuits** [comment]
fopen returns NULL in case of failures, and set 'errno' as the corresponding error.
(See this link: https://man7.org/linux/man-pages/man3/fopen.3.html)
You can check errno to find a cause of the failure.
(See this link: https://man7.org/linux/man-pages/man3/errno.3.html)
ex)
#include <errno.h>
...
void __coverage__(~, ~)
{ 
...
f = fopen(~~~);
if (f == NULL)
    printf("errno: %s\n",strerror(errno));
...
} 
(see this link(strerror(errno): https://man7.org/linux/man-pages/man3/strerror.3.html)

**Comment by adilb99** [comment]
@RiceBiscuits  Thanks a lot, turns out it's just a `Permission denied` error. Launching with `sudo` fixed it for me.
